packages/agreement-readmodel-writer/test/agreement-readmodel-writer.V2.integration.test.ts

```diff
@@ -0,0 +1,187 @@
+      "AgreementSuspendedByProducer",
+      "AgreementSuspendedByConsumer",
+      "AgreementSuspendedByPlatform",
+      "AgreementRejected",
+      "AgreementArchivedByUpgrade",
+    ] as const;
+
+    const event = {
+      agreement: toAgreementV2(agreement),
+    };
+
+    for (const eventType of eventTypes) {
+      const message: AgreementEventEnvelopeV2 = {
+        event_version: 2,
+        sequence_num: 1,
+        stream_id: agreement.id,
+        version: 1,
+        type: eventType,
+        data: event,
+        log_date: new Date(),
+      };
+
+      await handleMessageV2(message, agreements);
+
+      const actualAgreement = await agreements.findOne({
+        "data.id": agreement.id.toString(),
+      });
+
+      const expectedAgreement = toReadModelAgreement(agreement);
+
+      expect(actualAgreement?.data).toMatchObject(expectedAgreement);
+
+      expect(spyUpdate).toHaveBeenCalled();
+      expect(spyDelete).not.toHaveBeenCalled();
+    }
+  });
+
+  it("should test all agreement consumer document events", async () => {
+    const spyUpdate = vi.spyOn(agreements, "updateOne");
+    const spyDelete = vi.spyOn(agreements, "deleteOne");
+
+    const agreement = getMockAgreement();
+
+    const eventTypesConsumerDocument = [
+      "AgreementConsumerDocumentAdded",
+      "AgreementConsumerDocumentRemoved",
+    ] as const;
+
+    for (const eventType of eventTypesConsumerDocument) {
+      const event = {
+        agreement: toAgreementV2(agreement),
+        documentId: generateId(),
+      };
+
+      const message: AgreementEventEnvelopeV2 = {
+        event_version: 2,
+        sequence_num: 1,
+        stream_id: agreement.id,
+        version: 1,
+        type: eventType,
+        data: event,
+        log_date: new Date(),
+      };
+
+      await handleMessageV2(message, agreements);
+
+      const actualAgreement = await agreements.findOne({
+        "data.id": agreement.id.toString(),
+      });
+
+      const expectedAgreement = toReadModelAgreement(agreement);
+
+      expect(actualAgreement?.data).toMatchObject(expectedAgreement);
+
+      expect(spyUpdate).toHaveBeenCalled();
+      expect(spyDelete).not.toHaveBeenCalled();
+    }
+  });
+
+  it("should test delete agreement", async () => {
+    const spyUpdate = vi.spyOn(agreements, "updateOne");
+    const spyDelete = vi.spyOn(agreements, "deleteOne");
+
+    const agreement = getMockAgreement();
+
+    const eventType = "AgreementDeleted";
+    const event = {
+      agreement: toAgreementV2(agreement),
+    };
+
+    const message: AgreementEventEnvelopeV2 = {
+      event_version: 2,
+      sequence_num: 1,
+      stream_id: agreement.id,
+      version: 1,
+      type: eventType,
+      data: event,
+      log_date: new Date(),
+    };
+
+    await handleMessageV2(message, agreements);
+
+    const actualAgreement = await agreements.findOne({
+      "data.id": agreement.id.toString(),
+    });
+
+    expect(actualAgreement).toBeNull();
```

Comment:
Here maybe it's better if we first create the Agreement on MongoDB and then proceed with the test

Reply:
✔️  same also for the previous tests, I would add a test that has an agreement already in MongoDB and the event that updates it, so that we test also the upsert behavior and not only the insert